pipeline/alignments/align.py

```diff
@@ -153,17 +192,105 @@ def write_priors(
+        eflomal.write_priors(priors_output, *priors_tuple)
+
+
+def remap(
+    src_path: str,
+    trg_path: str,
+    tok_src_path: str,
+    tok_trg_path: str,
+    aln_path: str,
+    output_aln_path: str,
+) -> None:
+    """
+    Remaps alignments that were calculated for Moses-tokenized corpus to whitespace-tokenized ones.
+    :param src_path: path to whitespace-tokenized sentences in source language
+    :param trg_path: path to whitespace-tokenized sentences in target language
+    :param tok_src_path: path to Moses-tokenized sentences in source language
+    :param tok_trg_path: path to Moses-tokenized sentences in target language
+    :param aln_path: path to the alignments calculated for Moses-tokenized corpus
+    :param output_aln_path: path to output alignments file remapped to whitespace-tokenized corpus
+    """
+    logger.info("Remapping alignments to whitespace tokenization")
+
+    with ExitStack() as stack:
+        pool = stack.enter_context(multiprocessing.Pool(processes=multiprocessing.cpu_count()))
+        # Buffering helps to minimize IO operations which speeds thing up significantly
+        output = stack.enter_context(open(output_aln_path, "w", buffering=500000))
+
+        lines = zip(
+            stack.enter_context(open(src_path)),
+            stack.enter_context(open(trg_path)),
+            stack.enter_context(open(tok_src_path)),
+            stack.enter_context(open(tok_trg_path)),
+            stack.enter_context(open(aln_path)),
+        )
+
+        # send lines to worker processes in chunks
+        for aln in tqdm(pool.imap(remap_line, lines, chunksize=10000), mininterval=10):
+            output.write(aln)
+
+
+def remap_line(params):
+    """
+    Remaps alignments for a single line in a corpus
+    """
+    src, trg, tok_src, tok_trg, aln = params
+    src_map = map_indices(tok_src, src)
+    trg_map = map_indices(tok_trg, trg)
+
+    remapped_aln = []
+    for pair in aln.split():
+        src_idx, trg_idx = map(int, pair.split("-"))
+        new_pair = (src_map[src_idx], trg_map[trg_idx])
+        if new_pair not in remapped_aln:
+            remapped_aln.append(new_pair)
+
+    return " ".join([f"{idx1}-{idx2}" for idx1, idx2 in remapped_aln]) + "\n"
+
+
+def map_indices(tok_sentence: str, orig_sentence: str) -> Dict[int, int]:
+    """
+    Map token indices from tokenized sentence to original sentence.
+    :param tok_sentence: tokenized sentence
+    :param orig_sentence: original sentence
+    :return: Dictionary of indices that maps tokenized words to original words
+    """
+    tok_words = tok_sentence.split()
+    orig_words = orig_sentence.split()
+    tok_to_orig_indices = {}
+    orig_idx = 0
+    tok_idx = 0
+
+    # For 'Hello, world!'
+    # Map ['Hello', ',', 'world', '!'] [0, 1, 2, 3]
+    # To ['Hello,', 'world!'] [0, 1]
+    # tok -> orig: {0: 0, 1: 0, 2: 1, 3: 1}
+
+    while orig_idx < len(orig_words):
```

Comment:
Thought: If this logic were extended a bit, we could remap arbitrary tokenizations, even if the boundaries aren't as well defined as here.

Reply:
Yes, it's what OpusTrainer does with its Retokenize modifier. I considered resuing its code but it didn't fit what we need to do here.